fontc/src/main.rs

```diff
@@ -132,66 +138,75 @@ fn ir_source(source: &Path) -> Result<Box<dyn Source>, Error> {
 fn finish_successfully(context: ChangeDetector) -> Result<(), Error> {
     let current_sources =
         serde_yaml::to_string(&context.current_inputs).map_err(Error::YamlSerError)?;
-    fs::write(context.paths.ir_input_file(), current_sources).map_err(Error::IoError)
+    fs::write(context.ir_paths.ir_input_file(), current_sources).map_err(Error::IoError)
 }
 
-fn static_metadata_change(current_inputs: &Input, prev_inputs: &Input) -> bool {
-    current_inputs.static_metadata != prev_inputs.static_metadata
-}
+impl ChangeDetector {
+    fn static_metadata_ir_change(&self) -> bool {
+        self.current_inputs.static_metadata != self.prev_inputs.static_metadata
+            || !self
+                .ir_paths
+                .target_file(&FeWorkIdentifier::StaticMetadata)
+                .is_file()
+    }
 
-fn glyphs_changed(context: &ChangeDetector) -> IndexSet<&str> {
-    if static_metadata_change(&context.current_inputs, &context.prev_inputs) {
-        return context
-            .current_inputs
+    fn glyphs_changed(&self) -> IndexSet<&str> {
+        if self.static_metadata_ir_change() {
+            return self
+                .current_inputs
+                .glyphs
+                .keys()
+                .map(|e| e.as_str())
+                .collect();
+        }
+        self.current_inputs
+            .glyphs
+            .iter()
+            .filter(
+                |(glyph_name, curr_state)| match self.prev_inputs.glyphs.get(*glyph_name) {
+                    Some(prev_state) => {
+                        // If the input changed or the output doesn't exist a rebuild is probably in order
+                        prev_state != *curr_state
+                            || !self
+                                .ir_paths
+                                .target_file(&FeWorkIdentifier::Glyph(glyph_name.to_string()))
+                                .exists()
+                    }
+                    None => true,
+                },
+            )
+            .map(|(glyph_name, _)| glyph_name.as_str())
```

Comment:
this might be *slightly* easier expressed using `filter_map` 🤷

Reply:
Oh yes indeed. I'm used to having filter, I'm used to having map, apparently I'm not quite yet used to having filter_map

Reply:
So updated.